src/coreclr/jit/codegenarm64.cpp

```diff
@@ -3280,7 +3280,7 @@ void CodeGen::genCodeForStoreInd(GenTreeStoreInd* tree)
             else
             {
                 // issue a full memory barrier before a volatile StInd
-                instGen_MemoryBarrier();
+                instGen_MemoryBarrier(BARRIER_STORE_ONLY);
```

Comment:
What is the typical scenario when this branch is taken? When value is a struct?

Reply:
As I understand the purpose of this barrier is to have release semantics for storing into volatile variable that does not fit into a single register (otherwise stlr could be used). 

I think this needs a full barrier, since unlike `stlr`, `dmb ishst` does not wait for loads in progress.

Reply:
Basically, `ldar` can be replaced with `ldr; dmb ishld` , but there is no such equivalency between `stlr` and `dmb ishst; str` because ishst is too weak.

Reply:
From https://developer.arm.com/documentation/100941/0100/Barriers, trying to understand what you stated.

![image](https://user-images.githubusercontent.com/12488060/147993407-f4664533-c43a-45a0-be9b-2c21fafec207.png)

![image](https://user-images.githubusercontent.com/12488060/147993434-c5a54027-7d09-4f36-82ab-158a9961b06e.png)

Basically, for `ishst`, loads can still be reordered around barrier and hence it is weaker than the `ishld` where loads/stores need to wait till the barrier is complete.

If we change from full barrier to `ishst`, we might have a load that should have been completed but got reordered and might end up reading the wrong value (pre-updated value). Is my understanding correct?

Reply:
Yes, loads that appear after ishst, in program order, may speculatively happen ahead of the store.

Reply:
```cs
static volatile int x;
static volatile int y;

static int xx;
static int yy;

---- one thread:
x = 42;
yy = y;

--- another thread:
y = 42;
xx = x;

```

Can both `xx` and `yy` end up 0 ?

Reply:
Makes sense. I will revert the change related to `shst` then.